### PR TITLE
feat: add igv cnv files for WGS

### LINKED ISSUE #1493: [User Story] Delivery of additional CNV and BAF files to Caesar
### Need

As a researcher / clinician doing interpretation of CNVs from tumor normal matched WGS analyses I would like to have access to:
- the BAF plots loaded into GENS but in a format accessible by IGV (because when working from home GENS can be too slow)
- the CNV.somatic.*.ascat.copynumber.txt.gz file but from the normal sample which can be useful to get the uniparental disomy

See freshdesk ticket with request: https://scilifelab.freshdesk.com/a/tickets/1000758

### Suggested approach

**For Ascat:** I don't know if Ascat produces this file from the normal sample, and I can't see if it is because all outputs from Ascat are saved in a temporary folder which is deleted on completion. 
- [ ] Verify that Ascat produces the file for the normal sample
- [ ] If above is true then we just need to save this and store + deliver it

**For BAF tracks in IGV:**: 
Maybe we can expand the functionality of the GENS script to have an IGV option where the BAF folder does not have the GENS specific zoom levels.  Then we can just add an additional rule, baf_to_igv or something, and use this script, then store an deliver. 

### Considered alternatives

_No response_

### Deviation

_No response_

### System requirements assessed

- [ ] Yes, I have reviewed the system requirements

### Requirements affected by this story

_No response_

### Risk assessment needed

- [ ] Needed
- [ ] Not needed

### Risk assessment

_No response_

### SOUPs

_No response_

### Can be closed when

_No response_

### Blockers

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mathiasbio**
## Description

Add BAF and Log2 files from GATK to be stored in housekeeper and delivered to Caesar. See linked issue.

Added binning of 100-bp windows from GATK collect read-counts to reduce computational load when viewing in IGV (merging 5 bins by default) 

#### Added

- gnomad af 5 baf bedgraph for viewing in IGV
- denoised gatk readcounts log2 bedgraph for viewing in IGV

## Documentation

- [ ] N/A
- [x] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - docs/balsamic_sv_cnv.rst

### Feature Tests

- [x] Test: WGS TN case finished successfully and produces the new files
- [ ] Test: New files from WGS TN case can be stored in housekeeper (with associated hermes update) with the correct delivery tags:
    - [Screenshot]

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [ ] N/A
    - [x] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### Panel of Normal specific criteria

- [ ] The PR includes the addition of a new Panel of Normals
- [ ] The samples have been verified to adhere to the sample selection criteria on [Atlas PoN creation instructions for Balsamic](https://atlas.scilifelab.se/infrastructure/BALSAMIC/balsamic_pon/#procedure)

### User Changes

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

### Validation criteria

Validation criteria to be added to validation report PR: [LINK-TO-VALIDATION-REPORT-PR from the [validations repository](https://github.com/Clinical-Genomics/validations/)]

**Version specific criteria**

- Text here or N/A

> [!IMPORTANT]
> One of the below checkboxes for validation need to be checked 

- [ ] Added version specific validation criteria to validation report
- [ ] Changes validated in standard sections: [validation-section]
- [ ] Validation criteria not necessary

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [ ] Provided a comprehensive description of the PR.
    - [ ] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [ ] Verified and updated documentation if necessary.
- **Validation criteria**
    - [ ] Completed the validation criteria section of the template.
- **Tests**
    - [ ] Described and tested the functionality addressed in the PR.
    - [ ] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [ ] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Validation criteria**
    - [ ] The author has completed the validation criteria section of the template
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

**Review by fevac — COMMENTED** [review]
Oh sorry! I accidentally started to review this thinking it was the other PR. So I won't continue for now

File: BALSAMIC/assets/scripts/igv_baf_bedgraph.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+def open_output(path: str | Path) -> TextIO:
+    """Open output file; '-' means stdout."""
+    p = str(path)
+    if p == "-":
+        return sys.stdout
+    return open(p, "w", encoding="utf-8")
+
+
+@dataclass
```
[the remark this answers is not in the document]
Comment (mathiasbio) [reply BALSAMIC/assets/scripts/igv_baf_bedgraph.py]: I removed it and just replaced with an fstring

File: BALSAMIC/assets/scripts/igv_baf_bedgraph.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        af = 1.0
+    elif alt_sum == 0:
+        af = 0.0
+    else:
+        af = round(alt_sum / dp_val, 5)
+
+    start = int(variant.POS)
+    end = start + 1
+
+    return f"{variant.CHROM}\t{start}\t{end}\t{af}\n"
```
Comment (beatrizsavinhas) [inline BALSAMIC/assets/scripts/igv_baf_bedgraph.py]: I think it's best to avoid ambiguous return types from functions so perhaps it would be nice to raise an error here where you would return `None` and then add a try statement when the function is called.

Feel free to not do this since it will work either way, so it's just a matter of readability. If you like this design, do change the error message as you see fit since I wasn't sure of what to write. 😅 

```suggestion
def variant_to_record(variant) -> str:
    """
    Convert a cyvcf2 Variant to a bedGraph line.

    Assumptions:
      - exactly one sample
      - FORMAT/AD exists and has at least ref+alt counts
      - FORMAT/DP exists
    """
    ad = variant.format("AD")
    if not ad:
        raise Exception("No variant with format "AD" found.")

    counts = ad[0]
    alt_sum = int(np.sum(counts[1:]))

    dp = variant.format("DP")
    dp_val = int(np.asarray(dp).reshape(-1)[0])

    # Compute AF safely
    if dp_val <= 0:
        af = 1.0
    elif alt_sum == 0:
        af = 0.0
    else:
        af = round(alt_sum / dp_val, 5)

    start = int(variant.POS)
    end = start + 1

    return f"{variant.CHROM}\t{start}\t{end}\t{af}\n"
```

File: BALSAMIC/assets/scripts/igv_baf_bedgraph.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    with closing(VCF(str(vcf_path))) as vcf, open_output(bedgraph_path) as fout:
+        if track_name:
+            fout.write(f'track type=bedGraph name="{track_name}"\n')
+
+        for variant in vcf:
+            record = variant_to_record(variant)
+            if record == None:
+                continue
+            fout.write(record)
+            n_written += 1
```
Comment (beatrizsavinhas) [inline BALSAMIC/assets/scripts/igv_baf_bedgraph.py]: See [comment above on raising an exception for `variant_to_record`](https://github.com/Clinical-Genomics/BALSAMIC/pull/1649/changes#r2792016089).
```suggestion
			try:
            	record = variant_to_record(variant)
	            fout.write(record)
	            n_written += 1
            except:
                continue
```

File: BALSAMIC/constants/rules.py [hunk 4]
```diff
@@ -135,6 +135,7 @@
         ],
         "varcall": [
             "snakemake_rules/variant_calling/germline_wgs.rule",
+            "snakemake_rules/variant_calling/igv_files.rule",
```
Comment (beatrizsavinhas) [inline BALSAMIC/constants/rules.py]: The User story mentions only "tumor normal matched WGS analyses". Is this to be applied for TO too?

**Review by beatrizsavinhas — COMMENTED** [review]
I unfortunately don't fully understand the context around generating these files from reading the PR/user story description, so I can't really comment on logic or parameters used... 
But I added some comments on the code and suggestions for better readability!

If we meet to discuss the PR, maybe I can give a more complete and meaningful review! It would also be important to look at the test results.

File: BALSAMIC/assets/scripts/igv_baf_bedgraph.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    Convert a cyvcf2 Variant to a bedGraph line.
+
+    Assumptions:
+      - exactly one sample
+      - FORMAT/AD exists and has at least ref+alt counts
+      - FORMAT/DP exists
+    """
+    ad = variant.format("AD")
+    if ad is None:
+        return None
```
[the remark this answers is not in the document]
Comment (mathiasbio) [reply BALSAMIC/assets/scripts/igv_baf_bedgraph.py]: This function only takes in 1 variant, and the input VCF is quite strange so I expect quite a few of these unusual variants (variantcaller forces calls on gnomad positions). It doesn't matter so much if some variants are ignored, so I don't want the script to fail on this. But thank you!

File: BALSAMIC/assets/scripts/igv_baf_bedgraph.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    end = start + 1
+
+    return f"{variant.CHROM}\t{start}\t{end}\t{af}\n"
+
+
+def convert_vcf_to_bedgraph(
+    vcf_path: str | Path,
+    bedgraph_path: str | Path,
+    track_name: Optional[str] = None,
+) -> int:
```
[the remark this answers is not in the document]
Comment (mathiasbio) [reply BALSAMIC/assets/scripts/igv_baf_bedgraph.py]: I don't know, I think it's kind of neat to have the final print log at the very end of the script. Either way I don't have strong opinions on this, but in the interest of saving time I'd prefer not to change 😅
